ISPN-13986 Avoiding hybrid query if count targets an aggregable field

fax4ever opened the pull request:
https://issues.redhat.com/browse/ISPN-13986

Based on https://github.com/infinispan/infinispan/pull/11410

fax4ever commented:
@yrodiere could you review it please?

yrodiere reviewed (commented):
I may be missing something so I won't request a change, but I really don't understand how you would achieve this sort of aggregations with Hibernate Search... see below.

I suspect you need to limit yourself to `count(*)`/`count(<key>)`.

Otherwise, you really do need something like https://hibernate.atlassian.net/browse/HSEARCH-3661 , but for the Lucene backend. Which, as we discussed, is certainly possible, but not at all trivial, because the current implementation of Lucene aggregations is based on a Lucene "faceting" feature that basically only supports simple document counts by value, and no aggregation nesting at all.

fax4ever commented:
I pushed a changes in order to support the feature also if the query is a broadcast query.

fax4ever commented:
> I may be missing something so I won't request a change, but I really don't understand how you would achieve this sort of aggregations with Hibernate Search... see below.
> 
> I suspect you need to limit yourself to `count(*)`/`count(<key>)`.
> 
> Otherwise, you really do need something like https://hibernate.atlassian.net/browse/HSEARCH-3661 , but for the Lucene backend. Which, as we discussed, is certainly possible, but not at all trivial, because the current implementation of Lucene aggregations is based on a Lucene "faceting" feature that basically only supports simple document counts by value, and no aggregation nesting at all.

I started this contribution from the query from a user that does something like:

``` sql
select status, count(code) from org.infinispan.query.model.Sale where day = :day group by status order by status
```

@aleoncini could you please confirm it?

where it expressed a field name in the aggregation function. So I interpreted it to count the entity only if the code is not null (e.g.,  an exist predicate).
At this very moment it is not even possible to express a count on alias or on asterisk, to count the entity in any case and that's is reason because I opened the follow up => https://github.com/infinispan/infinispan/pull/11412.

yrodiere commented on query/src/main/java/org/infinispan/query/dsl/embedded/impl/SearchQueryMaker.java:
```diff
@@ -115,17 +118,48 @@ public SearchQueryParsingResult transform(IckleParsingResult<TypeMetadata> parsi
+         return null;
+      }
+
+      SearchAggregation<? extends Map<T, Long>> searchAggregation = scope.aggregation().terms()
+            .field(groupBy[0].asStringPathWithoutAlias(), projectedType).toAggregation();
```
You seem to be ignoring the sort here... You should probably use [`.orderByTermAscending()`](https://docs.jboss.org/hibernate/stable/search/reference/en-US/html_single/#search-dsl-aggregation-terms-order) depending on the query's `order by` clause?

Also, be aware that [the aggregation will return at most `100` entries by default](https://docs.jboss.org/hibernate/stable/search/reference/en-US/html_single/#search-dsl-aggregation-terms-maxtermcount). You might want to tune this or at least document it in Infinispan.

fax4ever replied:
> Also, be aware that [the aggregation will return at most 100 entries by default](https://docs.jboss.org/hibernate/stable/search/reference/en-US/html_single/#search-dsl-aggregation-terms-maxtermcount). You might want to tune this or at least document it in Infinispan.

I think I've already tuned it.

> You seem to be ignoring the sort here... You should probably use [.orderByTermAscending()](https://docs.jboss.org/hibernate/stable/search/reference/en-US/html_single/#search-dsl-aggregation-terms-order) depending on the query's order by clause?

I used the sort, it seems to be working

yrodiere replied:
Then that's one more things I don't understand, because the very code I commented on doesn't set either option:

```
      SearchAggregation<? extends Map<T, Long>> searchAggregation = scope.aggregation().terms()
            .field(groupBy[0].asStringPathWithoutAlias(), projectedType).toAggregation();
```

aleoncini replied:
> I started this contribution from the query from a user that does something like:
> 
> select status, count(code) from org.infinispan.query.model.Sale where day = :day group by status order by status
> @aleoncini could you please confirm it?

yes confirm, the query well represents the use case of the customer. HTH

yrodiere reviewed (commented):
> So I interpreted it to count the entity only if the code is not null (e.g., an exist predicate).

Ah, right, this seems to be the definition of `count(column)` in SQL, and I suppose in Ickle too. I mixed it up with `count(DISTINCT code)`, which is not what you were after.

This should be more manageable than counting the number of `code` values.

So, I see you insert a predicate into the Hibernate Search query to ignore `null` codes... While in most cases it should work, I think this will lead to incorrect results if a given `status` is only ever present with a `null` code: in that case you won't get a row for that status, even though you'd have expected one (with count == 0).

You could use [`.minDocumentCount(0)`](https://docs.jboss.org/hibernate/stable/search/reference/en-US/html_single/#search-dsl-aggregation-terms-mindocumentcount) in the aggregation definition, but then this means you'll get a row even for statuses for which no document matches the user-provided predicate (`where day = :day`).

I'm not sure I have a solution for you... :thinking:

fax4ever commented on query/src/main/java/org/infinispan/query/dsl/embedded/impl/QueryAnalyzer.java:
```diff
@@ -0,0 +1,90 @@
+
+      // we need at least one aggregation and one projection (the latter targeting the group by path)
+      for (PropertyPath projectedPath : projectedPaths) {
+         if (projectedPath instanceof AggregationPropertyPath) {
+            aggregation = (AggregationPropertyPath) projectedPath;
+
+            // only count aggregation function is currently supported by Hibernate Search
+            if (!AggregationFunction.COUNT.equals(aggregation.getAggregationFunction())) {
+               return false;
+            }
+            boolean aggregationTargetsAProperty = propertyHelper.hasProperty(parsingResult.getTargetEntityMetadata(), aggregation.asArrayPath());
+            if (aggregationTargetsAProperty) {
+               if (!metadata.isSearchable(aggregation.asArrayPath())) {
+                  return false;
+               }
+            }
```
> the path targeted by the aggregation. I'd expect that select count(code) group by status wouldn't work, since Hibernate Search aggregations count the number of documents, not the number of values... ? See my other comment about this.

If you see the test, you can see that only the entities having a code != null are counted. So it seems to be working.

fax4ever replied:
About that:

> the number of aggregations. What if there are two (valid) aggregations, e.g. select status, count(employer), count(code) group by status order by status? It seems you will just erase count(employer)

With this contribution if there are two (valid) aggregations, simply it will be a fallback to the current (non optimal) implementation. I wanted to keep the scope of this issue limited to improving the performances of one aggregation only. But feel free to open an issue as a follow up.

yrodiere replied:
> With this contribution if there are two (valid) aggregations, simply the execution will be fall-backed to the current (non optimal) implementation

Ok, I didn't see that. I see a check for `groupBys.length != 1` but not for "there is at most 1 aggregation". I must have missed the test too.

> I wanted to keep the scope of this issue limited to improving the performances of one aggregation only

Sure, makes sense.

> But feel free to open an issue as a follow up

Nope, I'm fine ;)

fax4ever commented on query/src/main/java/org/infinispan/query/dsl/embedded/impl/SearchQueryMaker.java:
```diff
@@ -115,17 +118,48 @@ public SearchQueryParsingResult transform(IckleParsingResult<TypeMetadata> parsi
+
+      for (int i=0; i<parsingResult.getProjectedPaths().length; i++) {
+         PropertyPath projectedPath = parsingResult.getProjectedPaths()[i];
+         if (projectedPath instanceof AggregationPropertyPath) {
+            if (projectedType != null) {
+               displayGroupFirst = true;
+            }
+            aggregationPropertyPath = (AggregationPropertyPath) projectedPath;
+         } else if (Arrays.equals(groupBy[0].asArrayPath(), projectedPath.asArrayPath())) {
+            projectedType = (Class<T>) parsingResult.getProjectedTypes()[i];
+         }
```
Yep. Multiple counts will be fall-backed to the current implementation.

fax4ever commented on client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/aggregation/RemoteQueryAggregationCountTest.java:
```diff
@@ -0,0 +1,59 @@
+
+      Query<Object[]> query;
+
+      query = remoteCache.query("select status, count(code) from Sale where day = :day group by status order by status");
```
> Is there some sort of post-processing I'm missing?

I don't. If I did it, the query wouldn't be fast enough.
What I did is to add the condition "code != null" in the predicate.

According to test it looks working to me.

fax4ever commented:
@yrodiere thanks for the review!

yrodiere commented:
FWIW I won't go through the conversations again but here are a few cases that might be problematic. 
I may be wrong, but I couldn't find tests for those cases, so... who knows.

Should be easy to fix:

* A query with two `count` on different fields, e.g. `select status, count(employer), count(code) group by status order by status`, will ignore the first `count()`. AFAICS this is not tested. See also https://github.com/infinispan/infinispan/pull/11410#discussion_r1366541633
* Results won't be sorted by term, but by document count. This might not be visible in tests because all results have the same document count, and you probably end up with the index order, which happens to match the term order, but only by chance. See https://github.com/infinispan/infinispan/pull/11410#discussion_r1369947896
* A query that is supposed to return more than 100 results will only return the first 100 results. This might not be visible in tests because there are fewer than 100 results. See also https://github.com/infinispan/infinispan/pull/11410#discussion_r1369947896 

I don't know how to fix:

* For a query like `select status, count(code) from org.infinispan.query.model.Sale where day = :day group by status order by status`, if the only document with a given status has a `null` code, the results will be missing the expected row (`status = thestatus, count = 0`). See also https://github.com/infinispan/infinispan/pull/11410#pullrequestreview-1694511801

fax4ever commented:
Hi @yrodiere, thank you for the review.
100 results would be a super critical thing to fix. Are you sure that it does not work?

> A query with two count on different fields, e.g. select status, count(employer), count(code) group by status order by status, will ignore the first count(). AFAICS this is not tested. See also https://github.com/infinispan/infinispan/pull/11410#discussion_r1366541633

I think I fixed it. Now if a query with 2 aggregations are used, the system will use the default (preexisting) non-indexed implementations.

> Results won't be sorted by term, but by document count. This might not be visible in tests because all results have the same document count, and you probably end up with the index order, which happens to match the term order, but only by chance. See https://github.com/infinispan/infinispan/pull/11410#discussion_r1369947896

We already re-sorted the result in the `AggregationDistributedIterator`. Now I fixed a local indexed query to do the same.

> A query that is supposed to return more than 100 results will only return the first 100 results. This might not be visible in tests because there are fewer than 100 results. See also https://github.com/infinispan/infinispan/pull/11410#discussion_r1369947896

This would be a serious issue. I increased the volumes of the indexed entities in tests. For instance the query without the where returns correctly a total count 10_000, that is > 100 I think. Isn't it?

> For a query like select status, count(code) from org.infinispan.query.model.Sale where day = :day group by status order by status, if the only document with a given status has a null code, the results will be missing the expected row (status = thestatus, count = 0). See also https://github.com/infinispan/infinispan/pull/11410#pullrequestreview-1694511801

Having no row in case of count 0 is something we can accept. I think there is no reason to change it now.

yrodiere commented:
> I think I fixed it. Now if a query with 2 aggregations are used, the system will use the default (preexisting) non-indexed implementations.

Makes sense.

> We already re-sorted the result in the AggregationDistributedIterator. Now I fixed a local indexed query to do the same.

Makes sense.

> This would be a serious issue. I increased the volumes of the indexed entities in tests. For instance the query without the where returns correctly a total count 10_000, that is > 100 I think. Isn't it?

No, it is not. The returned count is irrelevant. I'm talking about an aggregation that returns more than 100 buckets, i.e. with more than 100 different terms for the field you're grouping by.

I'll point you to the documentation again: https://docs.jboss.org/hibernate/stable/search/reference/en-US/html_single/#search-dsl-aggregation-terms-maxtermcount

You might want to look at Hibernate Search tests if it's still unclear:

* https://github.com/hibernate/hibernate-search/blob/6354c347a466be9f8a04d4bf985736b3bcdf68cc/integrationtest/backend/tck/src/main/java/org/hibernate/search/integrationtest/backend/tck/search/aggregation/TermsAggregationSpecificsIT.java#L458-L486
* https://github.com/hibernate/hibernate-search/blob/6354c347a466be9f8a04d4bf985736b3bcdf68cc/integrationtest/backend/tck/src/main/java/org/hibernate/search/integrationtest/backend/tck/search/aggregation/TermsAggregationSpecificsIT.java#L600-L624

> Having no row in case of count 0 is something we can accept. I think there is no reason to change it now.

And more importantly there is no _way_ to fix it :D

tristantarrant commented:
Merged via https://github.com/infinispan/infinispan/pull/11412
